src/Compilers/CSharp/Portable/Syntax/CSharpPragmaWarningStateMap.cs

```diff
@@ -17,85 +41,135 @@ internal class CSharpPragmaWarningStateMap : AbstractWarningStateMap
+                            else if (currentErrorCode.Kind() == SyntaxKind.IdentifierName)
+                            {
+                                errorId = ((IdentifierNameSyntax)currentErrorCode).Identifier.ValueText;
+                            }
+
+                            if (!string.IsNullOrWhiteSpace(errorId))
+                            {
+                                // Update the state of this error code with the current directive state
+                                accumulatedSpecificWarningState = accumulatedSpecificWarningState.SetItem(errorId, directiveState);
+                            }
+                        }
+                    }
                 }
                 else
                 {
-                    // Compute warning numbers from the current directive's codes
-                    for (int x = 0; x < currentDirective.ErrorCodes.Count; x++)
-                    {
-                        var currentErrorCode = currentDirective.ErrorCodes[x];
-                        if (currentErrorCode.IsMissing || currentErrorCode.ContainsDiagnostics)
-                            continue;
+                    var currentNullableDirective = (NullableDirectiveTriviaSyntax)currentDirective;
+                    PragmaWarningState directiveState;
 
-                        var errorId = string.Empty;
-                        if (currentErrorCode.Kind() == SyntaxKind.NumericLiteralExpression)
-                        {
-                            var token = ((LiteralExpressionSyntax)currentErrorCode).Token;
-                            errorId = MessageProvider.Instance.GetIdForErrorCode((int)token.Value);
-                        }
-                        else if (currentErrorCode.Kind() == SyntaxKind.IdentifierName)
-                        {
-                            errorId = ((IdentifierNameSyntax)currentErrorCode).Identifier.ValueText;
-                        }
+                    switch (currentNullableDirective.SettingToken.Kind())
+                    {
+                        case SyntaxKind.DisableKeyword:
+                            directiveState = PragmaWarningState.Disabled;
+                            break;
+                        case SyntaxKind.EnableKeyword:
+                            directiveState = PragmaWarningState.Enabled;
+                            break;
+                        default:
+                            throw ExceptionUtilities.UnexpectedValue(currentNullableDirective.SettingToken.Kind());
+                    }
 
-                        if (!string.IsNullOrWhiteSpace(errorId))
-                        {
-                            // Update the state of this error code with the current directive state
-                            accumulatedSpecificWarningState = accumulatedSpecificWarningState.SetItem(errorId, directiveState);
-                        }
+                    foreach (string id in ErrorFacts.NullableFlowAnalysisWarnings)
```

Comment:
>foreach [](start = 20, length = 7)

Perhaps use `ImmutableDictionary<K, V>.SetItems()` to reduce the number of allocations.